altair/typing.py

```diff
@@ -1,63 +1,5 @@
 from __future__ import annotations
 
-import sys
-from typing import Any, Mapping, Union
-from typing_extensions import TypedDict
+__all__ = ["EncodeKwds"]
 
-if sys.version_info >= (3, 10):
-    from typing import TypeAlias
-else:
-    from typing_extensions import TypeAlias
-
-ChannelType: TypeAlias = Union[str, Mapping[str, Any], Any]
```

Comment:
@binste I know I removed `ChannelType` here, but `altair.typing` could define the precise types for each of the positional-args:

https://github.com/pola-rs/polars/pull/17995#issuecomment-2263609625

![image](https://github.com/user-attachments/assets/0a1dcfd7-c0b9-4c60-ba3e-4067fa73316a)

These did vary between each method, but I think this would be 6 unique channel type aliases total:

- https://github.com/MarcoGorelli/polars/blob/db6d8f76782aa02bb72b923cf46e2f638ede6bef/py-polars/polars/dataframe/plotting.py#L30
- https://github.com/MarcoGorelli/polars/blob/db6d8f76782aa02bb72b923cf46e2f638ede6bef/py-polars/polars/dataframe/plotting.py#L92
- https://github.com/MarcoGorelli/polars/blob/db6d8f76782aa02bb72b923cf46e2f638ede6bef/py-polars/polars/dataframe/plotting.py#L161


Unsure if there would be more needed for the yet-unwritten methods

Reply:
You already got very far! Great to see that we're almost there 😄 

I'm going through it now but I'm also wondering if we maybe just want to specify, for each channel, the types separately. `ChannelX`, `ChannelY`, etc. works for me regarding naming. So basically what's there now for `EncodingKwgs` but as individual type aliases.

I don't know if we then even need a `TypedDict`, what do you think? `Unpack` supports the usage suggested in https://github.com/pola-rs/polars/pull/17995#issuecomment-2263609625 only for Python >= 3.12. And I think the plot methods should then only accept `**kwargs: Unpack[EncodeKwgs]` as else the arguments which are explicitly defined are defined twice. See [typing docs](https://typing.readthedocs.io/en/latest/spec/callables.html#keyword-collisions).

Downside would be that in Polars, users only get autocomplete and type checking for the channels which are explicitly defined but for simple exploratory charts, this could be enough?

Reply:
> You already got very far! Great to see that we're almost there 😄

Thanks @binste for such a thorough response!

> And I think the plot methods should then only accept `**kwargs: Unpack[EncodeKwds]` as else the arguments which are explicitly defined are defined twice. See [typing docs](https://typing.readthedocs.io/en/latest/spec/callables.html#keyword-collisions).

I was trying to demonstrate the correct way to use `Unpack` there, to avoid that issue (the second one):

```py

def foo(name, **kwargs: Unpack[Movie]) -> None: ...     # WRONG! "name" will always bind to the first parameter.

def foo(name, /, **kwargs: Unpack[Movie]) -> None: ...  # OK! "name" is a positional-only parameter,
           # ^^^                                        # so **kwargs can contain a "name" keyword.
```

<details>
<summary>Long screenshot</summary>

![image](https://github.com/user-attachments/assets/382426a1-451e-4b52-bf59-df0cff2e2c1b)

</details>


> I don't know if we then even need a `TypedDict`, what do you think? `Unpack` supports the usage suggested in [pola-rs/polars#17995 (comment)](https://github.com/pola-rs/polars/pull/17995#issuecomment-2263609625) only for Python >= 3.12. 

I'm not sure there's an issue with [`typing_extensions.Unpack`](https://typing-extensions.readthedocs.io/en/latest/#typing_extensions.Unpack) which I used unconditionally there?
Usually `ruff` or `pylance` will warn me when I do something that isn't allowed in `py3.8`

![image](https://github.com/user-attachments/assets/5dcedff0-2ba6-4758-95c8-adedac5912f9)

---
@binste 
> Downside would be that in Polars, users only get autocomplete and type checking for the channels which are explicitly defined but for simple exploratory charts, this could be enough?

https://github.com/pola-rs/polars/pull/17995#issuecomment-2263609625
@dangotbanned 
> You could then repeat the `/, **kwargs: Unpack[EncodeKwds]` for the other methods - maybe changing the positional-only ones if needed

This was essentially the compromise of that, if that makes sense?
You have your common ones that can be called *without* kwargs, but more options *with* kwargs

Reply:
Yes, you're absolutely right! Thanks for the detailed explanation. Lost a bit the overview there but this brought me back on track :)

Just tested `/, **kwargs: Unpack[EncodeKwds]` with `typing_extensions.Unpack` on Python 3.11 and works as expected. 

Reviewing the rest now.

Reply:
@MarcoGorelli when you see this

The need for this solution is that `alt.Chart.encode` has alphabetically ordered parameters, whereas `pl.DataFrame.plot` has:
- pos-only (in a different order)
- *args
- **kwargs

Reply:
**Marking as unresolved purely so this is easier to find during review**

It is actually resolved unless https://github.com/vega/altair/pull/3515#discussion_r1703238386 turns up anything new